src/core/server/saved_objects/service/lib/internal_bulk_resolve.ts

```diff
@@ -0,0 +1,333 @@
+    ? await client.mget<SavedObjectsRawDocSource>(
+        { body: { docs: docsToBulkGet } },
+        { ignore: [404] }
+      )
+    : undefined;
+  // exit early if a 404 isn't from elasticsearch
+  if (
+    bulkGetResponse &&
+    isNotFoundFromUnsupportedServer({
+      statusCode: bulkGetResponse.statusCode,
+      headers: bulkGetResponse.headers,
+    })
+  ) {
+    throw SavedObjectsErrorHelpers.createGenericNotFoundEsUnavailableError();
+  }
+
+  let getResponseIndex = 0;
+  let aliasTargetIndex = 0;
+  const resolveCounter = new ResolveCounter();
+  const resolvedObjects = allObjects.map<SavedObjectsResolveResponse<T> | InternalBulkResolveError>(
+    (either) => {
+      if (isLeft(either)) {
+        return either.value;
+      }
+      const exactMatchDoc = bulkGetResponse?.body.docs[getResponseIndex++];
+      let aliasMatchDoc: MgetHit<SavedObjectsRawDocSource> | undefined;
+      const aliasTargetId = aliasTargetIds[aliasTargetIndex++];
+      if (aliasTargetId !== undefined) {
+        aliasMatchDoc = bulkGetResponse?.body.docs[getResponseIndex++];
+      }
+      const foundExactMatch =
+        // @ts-expect-error MultiGetHit._source is optional
+        exactMatchDoc.found && rawDocExistsInNamespace(registry, exactMatchDoc, namespace);
+      const foundAliasMatch =
+        // @ts-expect-error MultiGetHit._source is optional
+        aliasMatchDoc?.found && rawDocExistsInNamespace(registry, aliasMatchDoc, namespace);
+
+      const { type, id } = either.value;
+      let result: SavedObjectsResolveResponse<T> | null = null;
+      if (foundExactMatch && foundAliasMatch) {
+        result = {
+          // @ts-expect-error MultiGetHit._source is optional
+          saved_object: getSavedObjectFromSource(registry, type, id, exactMatchDoc),
+          outcome: 'conflict',
+          alias_target_id: aliasTargetId!,
+        };
+        resolveCounter.recordOutcome(REPOSITORY_RESOLVE_OUTCOME_STATS.CONFLICT);
+      } else if (foundExactMatch) {
+        result = {
+          // @ts-expect-error MultiGetHit._source is optional
+          saved_object: getSavedObjectFromSource(registry, type, id, exactMatchDoc),
+          outcome: 'exactMatch',
+        };
+        resolveCounter.recordOutcome(REPOSITORY_RESOLVE_OUTCOME_STATS.EXACT_MATCH);
+      } else if (foundAliasMatch) {
+        result = {
+          // @ts-expect-error MultiGetHit._source is optional
+          saved_object: getSavedObjectFromSource(registry, type, aliasTargetId!, aliasMatchDoc),
+          outcome: 'aliasMatch',
+          alias_target_id: aliasTargetId,
+        };
+        resolveCounter.recordOutcome(REPOSITORY_RESOLVE_OUTCOME_STATS.ALIAS_MATCH);
+      }
+
+      if (result !== null) {
+        return result;
+      }
+      resolveCounter.recordOutcome(REPOSITORY_RESOLVE_OUTCOME_STATS.NOT_FOUND);
+      return {
+        type,
+        id,
+        error: SavedObjectsErrorHelpers.createGenericNotFoundError(type, id),
+      };
```

Comment:
(Part 1 of 3)
I'm not thrilled with this, but I view it as an acceptable compromise.

Here is the _current_ behavior of `get` vs `resolve`:
* get
  * found: returns `{ type, id, attributes, references, ... }`
  * not found: throws error
* bulkGet
  * found: returns `[{ type, id, attributes, references, ... }]` 
  * not found: returns `[{ type, id, error }]`
* resolve
  * found: returns `{ saved_object: { type, id, attributes,references, ... }, outcome }`
  * not found: throws error
* bulkResolve - what to do about this? (see other comment below)

Note that the `SavedObject` type requires the `attributes` and `references` fields and has an optional `error` field that is only used for bulk responses. But the current behavior of `bulkGet` (and the other `bulk*` methods) doesn't return a valid `SavedObject` type, because only the `type`, `id`, and `error` fields are present, because the `SOR` uses `any` under the hood.

When writing the `internalBulkResolve` module, I wanted to be able to reuse it for both `SOR.resolve` and `SOR.bulkResolve` -- that meant that I needed to return the actual `DecoratedError` object so that the `SOR.resolve` method could find it and throw the error.

Reply:
> When writing the internalBulkResolve module, I wanted to be able to reuse it for both SOR.resolve and SOR.bulkResolve

That makes sense

> that meant that I needed to return the actual DecoratedError object so that the SOR.resolve method could find it and throw the error.

Can't we just use a higher level of abstraction by having `internalBulkResolve` returning the type of the error and letting the `resolve` and `bulkResolve` callers reformat/wrap them?

Reply:
> Can't we just use a higher level of abstraction by having `internalBulkResolve` returning the type of the error and letting the `resolve` and `bulkResolve` callers reformat/wrap them?

Well `resolve` needs to throw a `DecoratedError`, while `bulkResolve` needs to extract the `payload.output` field from it. So I figured the least common denominator is to have the `internalBulkResolve` return the `DecoratedError` itself.

If you think it's better to return the error _type_ instead that should be fine and I can make that change (we really just have two types, "unsupported" and "not found"). I'll wait til you and Rudolf get a chance to reply before doing anything else though.

Reply:
From a pure design standpoint, returning a real `Error` (or subclass) instance as an attribute of a return value is imho bad practice. Now, honestly, if that answers our current need, that's very fine with me.